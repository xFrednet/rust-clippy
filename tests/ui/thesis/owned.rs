//@rustc-env: CLIPPY_PETS_PRINT=1

struct Animal {
    legs: u32,
    heads: u32,
}

// Check arguments are correctly detected
#[warn(clippy::borrow_pats)]
fn take_one(_animal: Animal) {}

#[warn(clippy::borrow_pats)]
fn take_two(_animal_1: Animal, _animal_2: Animal) {}

fn take_pair((_animal_1, _animal_2): (Animal, Animal)) {}

#[warn(clippy::borrow_pats)]
fn pat_return_owned_arg(animal: Animal) -> Animal {
    animal
}

#[forbid(clippy::borrow_pats)]
// #[warn(clippy::borrow_pats)]
fn pat_maybe_return_owned_arg_1(a: String) -> String {
    if !a.is_empty() {
        return a;
    }

    "hey".to_string()
}

// #[forbid(clippy::borrow_pats)]
// #[warn(clippy::borrow_pats)]
fn pat_maybe_return_owned_arg_1_test(a: u32) -> u32 {
    if !a.is_power_of_two() {
        return a;
    }

    19
}

// #[forbid(clippy::borrow_pats)]
fn pat_maybe_return_owned_arg_2(a: String) -> String {
    let ret;
    if !a.is_empty() {
        ret = a;
    } else {
        ret = "hey".to_string();
        println!("{a:#?}");
    }
    ret
}

fn pat_copy_to_self(mut animal: Animal) {
    animal.heads = animal.legs;
}

fn main() {}
