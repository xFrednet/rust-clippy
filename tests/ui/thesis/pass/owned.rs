//@rustc-env: CLIPPY_PETS_PRINT=1
//@rustc-env: CLIPPY_PRINT_MIR=1

struct Dropper {
    duck: u32,
}

impl Drop for Dropper {
    fn drop(&mut self) {}
}

struct Animal {
    legs: u32,
    heads: u32,
}

// Check arguments are correctly detected
#[warn(clippy::borrow_pats)]
fn take_one(_animal: Animal) {}

#[warn(clippy::borrow_pats)]
fn take_two(_animal_1: Animal, _animal_2: Animal) {}

fn take_pair((_animal_1, _animal_2): (Animal, Animal)) {}

#[warn(clippy::borrow_pats)]
fn pat_return_owned_arg(animal: Animal) -> Animal {
    animal
}

#[warn(clippy::borrow_pats)]
fn pat_maybe_return_owned_arg_1(a: String) -> String {
    if !a.is_empty() {
        return a;
    }

    "hey".to_string()
}

#[warn(clippy::borrow_pats)]
fn pat_maybe_return_owned_arg_1_test(a: u32) -> u32 {
    if !a.is_power_of_two() {
        return a;
    }

    19
}

#[warn(clippy::borrow_pats)]
/// FIXME: The argument return is not yet detected both in `a`
fn pat_maybe_return_owned_arg_2(a: String) -> String {
    let ret;
    if !a.is_empty() {
        ret = a;
    } else {
        ret = "hey".to_string();
        println!("{a:#?}");
    }
    ret
}

#[warn(clippy::borrow_pats)]
fn pat_maybe_return_owned_arg_3(a: String) -> String {
    let ret = if !a.is_empty() { a } else { "hey".to_string() };
    ret
}

#[warn(clippy::borrow_pats)]
fn pub_dynamic_drop_1(animal: String, cond: bool) {
    if cond {
        // Move out of function
        std::mem::drop(animal);
    }

    magic()
}

#[warn(clippy::borrow_pats)]
fn conditional_overwrite(mut animal: String, cond: bool) {
    if cond {
        animal = "Ducks".to_string();
    }

    magic()
}

fn magic() {}

#[derive(Default)]
struct Example {
    owned_1: String,
    owned_2: String,
    copy_1: u32,
    copy_2: u32,
}

#[warn(clippy::borrow_pats)]
fn test_ctors() {
    let s1 = String::new();
    let _slice = (s1, 2);

    let s1 = String::new();
    let _array = [s1];

    let s1 = String::new();
    let _thing = Example {
        owned_1: s1,
        ..Example::default()
    };
}

#[warn(clippy::borrow_pats)]
fn main() {
    let dropper = Dropper { duck: 17 };
}
