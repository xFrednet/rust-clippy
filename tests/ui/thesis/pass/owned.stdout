# "take_one"
- _animal        : (Immutable, Owned    , Argument, NonCopy, NonDrop , UserDef  ) {}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {}

# "take_two"
- _animal_1      : (Immutable, Owned    , Argument, NonCopy, NonDrop , UserDef  ) {}
- _animal_2      : (Immutable, Owned    , Argument, NonCopy, NonDrop , UserDef  ) {}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {}

# "pat_return_owned_arg"
- animal         : (Immutable, Owned    , Argument, NonCopy, NonDrop , UserDef  ) {Moved, MovedToReturn}
- Body: Output(UserDef)  , NotConst , Safe  , Sync , Free       {}

# "pat_maybe_return_owned_arg_1"
- a              : (Immutable, Owned    , Argument, NonCopy, PartDrop, UserDef  ) {Moved, MovedToReturn, Borrow, ArgBorrow}
- Body: Output(UserDef)  , NotConst , Safe  , Sync , Free       {HasTempBorrow}

# "pat_maybe_return_owned_arg_1_test"
- a              : (Immutable, Owned    , Argument, Copy   , NonDrop , Primitive) {}
- Body: Output(Primitive), NotConst , Safe  , Sync , Free       {}

# "pat_maybe_return_owned_arg_2"
- a              : (Immutable, Owned    , Argument, NonCopy, PartDrop, UserDef  ) {DynamicDrop, Moved, MovedToVar, Borrow, ArgBorrow, ArgBorrowExtended, ConditionalMove}
- ret            : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Moved, MovedToReturn, ConditionalInit}
- Body: Output(UserDef)  , NotConst , Safe  , Sync , Free       {HasTempBorrow}

# "pat_maybe_return_owned_arg_3"
- a              : (Immutable, Owned    , Argument, NonCopy, PartDrop, UserDef  ) {DynamicDrop, Moved, MovedToVar, Borrow, ArgBorrow, ConditionalMove}
- ret            : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Moved, MovedToReturn, ConditionalInit}
- Body: Output(UserDef)  , NotConst , Safe  , Sync , Free       {HasTempBorrow}

# "pub_dynamic_drop_1"
- animal         : (Immutable, Owned    , Argument, NonCopy, PartDrop, UserDef  ) {DynamicDrop, Moved, MovedToFn, ManualDrop, ConditionalMove}
- cond           : (Immutable, Owned    , Argument, Copy   , NonDrop , Primitive) {}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {}

# "conditional_overwrite"
- animal         : (Mutable  , Owned    , Argument, NonCopy, PartDrop, UserDef  ) {Overwrite, ConditionalOverwride, ConditionalDrop, DropForReplacement}
- cond           : (Immutable, Owned    , Argument, Copy   , NonDrop , Primitive) {}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {}

# "test_ctors"
- s1             : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Moved, MovedToCtor}
- _slice         : (Immutable, Owned    , Local   , NonCopy, PartDrop, Tuple    ) {}
- s1             : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Moved, MovedToCtor}
- _array         : (Immutable, Owned    , Local   , NonCopy, PartDrop, Sequence ) {}
- s1             : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Moved, MovedToCtor}
- _thing         : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments}

# "main"
- dropper        : (Immutable, Owned    , Local   , NonCopy, SelfDrop, UserDef  ) {}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments}

